### ISSUE TITLE
Running several async filter functions using async

### ISSUE THREAD

**Issue by mercmobily** [caolan/async#808]
I have a bunch of functions with this signature:

    var plugins = [
      function f1( fileData, fileInfo, cb ){
        /* Manipulate fileData */
        fileData += "f1 ran! ";
        cb( null, fileData );
      },

      function f2( fileData, fileInfo, cb ){
        /* Manipulate fileData */
         fileData += "f2 ran! ";
        cb( null, fileData );
      }
    ];

I want to run all of them, in sequence, and get the end result.

At the moment, I am using `async.waterfall`:

    var async = require( 'async');

    var pluginsWithStarter = [ function( cb ){
      return cb( null, fileData, fileInfo );
    } ].concat( plugins );

    var fileData = "Some example data";
    var fileInfo = 

    async.waterfall( pluginsWithStarter, function( err, fileData, fileInfo){
      if( err ) return cb( err );

      console.log("RESULT: ");
      console.log( fileData );
      console.log( fileInfo );

      cb( null );
    });

Note that using this methods, plugins are forced to call the callback with _both_ arguments: `return cb( null, fileData, fileInfo );`.

Basically, I am appending a "starter" function to the waterfall, that will  run with the two initial parameters -- which all of other functions will process AND return.

Questions:

1) Is this a good way to use async.waterfall? I expected it to be able to take in itself the first parameters, like this: `async.waterfall( pluginsWithStarter, fileData, fileInfo, function( err, fileData, fileInfo){` However, no, you HAVE TO have a first function which takes no parameters and will then return the callback with several paramerters

2) This way I am forced to get every function to return BOTH values, which is a bit ugly. How could I change the code so that plugins only have to call the callback with one parameter?

**Comment by aearly** [comment]
1.  Yes, that's an okay way to use  waterfall -- having the first function be a simple function that sets up arguments.  The other option would be to do something like `plugins[0] = plugins[0].bind(null, fileData, fileInfo);` to partially apply the first function.

I'm also considering adding [`acomb.constant`](https://www.npmjs.com/package/acomb#constant) to async, to make creating that first function simpler:

```js
var pluginsWithStarter = [async.constant(fileData, fileInfo)].concat(plugins);
```

2.  I hear you about having to pass everything though the waterfall.  You can put variables in the parent closure scope to prevent having to manually pass everything through the waterfall.  You can also look into `async.auto` for managing many variables in an async workflow.  Its interface is a little bit weird, I will admit...

**Comment by mercmobily** [comment]
Thanks!
